Update the URL when changing mailer preview formats

jameskerr opened the pull request:
Added javascript to update the URL on mailer previews with the currently selected email format. Reloading the page now keeps you on your selected format rather than going back to the default html version.

This is done using javascript's history.pushState only if the developer's browser supports it.

jameskerr committed:
Update the URL when changing mailer preview formats

Added javascript to update the URL on mailer previews with the
currently selected email format. Reloading the page now keeps you on
your selected format rather than going back to the default html version.

arthurnn commented:
`pushState`, I hate messing with that =).
So, what is the problem you are trying to solve? Not sure if I fully understood the description, sorry.

thanks for the PR

jameskerr commented:
Hi @arthurnn, thanks for reviewing this!  Its my first time contributing to rails.

__The problem:__

When you create a mailer preview in rails, you can preview an email as the .html version or the .txt version.  You do this by using that select box in the header area of the mailer preview page.  When you load the mailer preview page, the select box defaults to the .html version.  

![screen shot 2015-10-05 at 10 20 27 am](https://cloud.githubusercontent.com/assets/3460638/10287925/c3a477f6-6b4a-11e5-89be-28e332d5aa28.png)

The annoyance begins when you want to work on the .txt version of the email.  Each time you make a change to the .txt email in your editor, you naturally reload your browser to see your changes.  However, each time you reload, it defaults to the .html version and you are forced to re-select the .txt version from the select box.

__The proposed solution:__

Each time the user selects the version they want to preview, we append that mime type to the url.  So, when I change the select box to show me the .txt version, I append .txt to the url.  When I then reload the browser, I am brought directly to the .txt version of the email.


After creating this pull request, I realized that `pushState` was not the way to go.  It was annoying for me to use.  I have added another commit that now uses `replaceState`.  I also noticed that the iframe was adding history when the `src` attribute was being changed.  So I updated that to use `iframe.contentWindow.location.replace(selected_format_url)`.  This does not add any history to the browser.

So now, developers can select their desired mailer preview format, work on it, reload the page, then click back one time and go to the page they were previously on (as opposed to clicking back through all the email formats they previously selected).

PS: Let me know if I am doing something stupid in git.  I feel like I didn't update my branch properly.

arthurnn commented:
@pixeltrix what do you think about this behaviour.

re: git, I think you mess it up =) . Try fetching latest rails/rails and rebase your branch against rails master.

pixeltrix commented:
:+1: to the behaviour but obviously the branch needs rebasing

jameskerr commented:
@pixeltrix @arthurnn I figured out how to rebase and squash the commits.  Let me know if there is any other change you would like me to make.  I have been using it on a sample app, and I am enjoying the new behavior already.  
![out](https://cloud.githubusercontent.com/assets/3460638/10318173/17df223a-6c1d-11e5-8f5f-b4f3c47af508.gif)

pixeltrix commented:
@jameskerr there are still merge conflicts preventing me from merging this - pull the latest master and then rebase against that, thanks

pixeltrix commented:
@jameskerr also the CHANGELOG entry should be at the top of the file

jameskerr commented:
@pixeltrix After some trial and error, I believe I have successfully rebased with master.  I put my change at the top of the changelog.

arthurnn commented:
This is great.. thanks @jameskerr . and good work. :heart:
